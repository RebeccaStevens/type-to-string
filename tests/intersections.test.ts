import { describe, it } from "vitest";

import { runTests } from "./helpers";

describe("Intersections", () => {
  describe("simple", () => {
    it.each([
      "type Test = Readonly<{ foo: string; }> & Readonly<{ bar: string; }>;",
      "type Test = readonly (number | string)[] & readonly (number | boolean)[];",
      "type Test = { foo: string; } & { bar: string; };",
      "type Test = Readonly<{ foo: string; } & { bar: string; }>;",
    ])("%s", (code) => {
      runTests(code);
    });
  });

  describe("same props", () => {
    it.each([
      "type Test = { readonly foo: ReadonlyArray<string>; } & { readonly foo: Array<string>; };",
      "type Test = ReadonlyArray<number> & { readonly 0: number; };",
    ])("%s", (code) => {
      runTests(code);
    });
  });
});
