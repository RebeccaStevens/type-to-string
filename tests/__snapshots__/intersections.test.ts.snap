// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Intersections > same props > type Test = { readonly foo: ReadonlyArray<string>; } & { readonly foo: Array<string>; }; 1`] = `
{
  "typeAlias": "Test",
  "typeReference": null,
  "typeToString": "Test",
}
`;

exports[`Intersections > same props > type Test = { readonly foo: ReadonlyArray<string>; } & { readonly foo: Array<string>; }; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "{ readonly foo: ReadonlyArray<string>; } & { readonly foo: Array<string>; }",
  "typeNodeToString": null,
  "typeReferenceNode": null,
}
`;

exports[`Intersections > same props > type Test = ReadonlyArray<number> & { readonly 0: number; }; 1`] = `
{
  "typeAlias": "Test",
  "typeReference": null,
  "typeToString": "Test",
}
`;

exports[`Intersections > same props > type Test = ReadonlyArray<number> & { readonly 0: number; }; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "ReadonlyArray<number> & { readonly 0: number; }",
  "typeNodeToString": null,
  "typeReferenceNode": null,
}
`;

exports[`Intersections > simple > type Test = { foo: string; } & { bar: string; }; 1`] = `
{
  "typeAlias": "Test",
  "typeReference": null,
  "typeToString": "Test",
}
`;

exports[`Intersections > simple > type Test = { foo: string; } & { bar: string; }; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "{ foo: string; } & { bar: string; }",
  "typeNodeToString": null,
  "typeReferenceNode": null,
}
`;

exports[`Intersections > simple > type Test = Readonly<{ foo: string; } & { bar: string; }>; 1`] = `
{
  "typeAlias": "Readonly<T>",
  "typeReference": null,
  "typeToString": "Readonly<{ foo: string; } & { bar: string; }>",
}
`;

exports[`Intersections > simple > type Test = Readonly<{ foo: string; } & { bar: string; }>; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "Readonly<{ foo: string; } & { bar: string; }>",
  "typeNodeToString": null,
  "typeReferenceNode": "Readonly<T>",
}
`;

exports[`Intersections > simple > type Test = Readonly<{ foo: string; }> & Readonly<{ bar: string; }>; 1`] = `
{
  "typeAlias": "Test",
  "typeReference": null,
  "typeToString": "Test",
}
`;

exports[`Intersections > simple > type Test = Readonly<{ foo: string; }> & Readonly<{ bar: string; }>; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "Readonly<{ foo: string; }> & Readonly<{ bar: string; }>",
  "typeNodeToString": null,
  "typeReferenceNode": null,
}
`;

exports[`Intersections > simple > type Test = readonly (number | string)[] & readonly (number | boolean)[]; 1`] = `
{
  "typeAlias": "Test",
  "typeReference": null,
  "typeToString": "Test",
}
`;

exports[`Intersections > simple > type Test = readonly (number | string)[] & readonly (number | boolean)[]; 2`] = `
{
  "typeNodeAlias": "Test",
  "typeNodeAsWritten": "readonly (number | string)[] & readonly (number | boolean)[]",
  "typeNodeToString": null,
  "typeReferenceNode": null,
}
`;
